src/main/java/io/divolte/server/kafka/KafkaFlusher.java

```diff
@@ -83,55 +79,84 @@ public ProcessingDirective process(final Queue<AvroRecordBuffer> batch) {
+            final ImmutableList<ProducerRecord<DivolteIdentifier,AvroRecordBuffer>> remaining = sendBatch(batch);
+            pendingMessages = remaining;
+            return remaining.isEmpty() ? CONTINUE : PAUSE;
+        } catch (final InterruptedException e) {
+            // This is painful. We don't know how much of the batch was stored, and how much wasn't.
+            // This should only occur during shutdown. I think we can assume that anything missed can be discarded.
+            logger.warn("Flushing interrupted. Not all messages in batch (size={}) may have been sent to Kafka.", batch.size());
+            pendingMessages = ImmutableList.of();
+            return CONTINUE;
         }
-        return result;
     }
 
-    private KeyedMessage<byte[], byte[]> buildMessage(final AvroRecordBuffer record) {
-        // Extract the AVRO record as a byte array.
-        // (There's no way to do this without copying the array.)
-        final ByteBuffer avroBuffer = record.getByteBuffer();
-        final byte[] avroBytes = new byte[avroBuffer.remaining()];
-        avroBuffer.get(avroBytes);
-        final String partyId = record.getPartyId().value;
-        return new KeyedMessage<>(topic, partyId.getBytes(StandardCharsets.UTF_8), partyId, avroBytes);
+    private ImmutableList<ProducerRecord<DivolteIdentifier,AvroRecordBuffer>> sendBatch(final List<ProducerRecord<DivolteIdentifier, AvroRecordBuffer>> batch) throws InterruptedException {
+        // First start sending the messages.
+        // (This will serialize them, determine the partition and then assign them to a per-partition buffer.)
+        final List<Future<RecordMetadata>> sendResults =
+                batch.stream()
+                     .map(producer::send)
+                     .collect(Collectors.toList());
+        // The producer will send the messages in the background. As of 0.8.x we can't
+        // flush, but have to wait for that to occur based on the producer configuration.
+        // (By default it will immediately flush, but users can override this.)
+
+        // When finished, each message can be in one of several states.
+        //  - Completed.
+        //  - An error occurred, but a retry may succeed.
+        //  - A fatal error occurred.
+        // (In addition, we can be interrupted due to shutdown.)
+        final ImmutableList.Builder<ProducerRecord<DivolteIdentifier, AvroRecordBuffer>> remaining = ImmutableList.builder();
+        final int batchSize = batch.size();
+        for (int i = 0; i < batchSize; ++i) {
+            final Future<RecordMetadata> result = sendResults.get(i);
+            try {
+                final RecordMetadata metadata = result.get();
+                if (logger.isTraceEnabled()) {
+                    final ProducerRecord<DivolteIdentifier, AvroRecordBuffer> record = batch.get(i);
+                    logger.trace("Finished sending event (partyId={}) to Kafka: topic/partition/offset = {}/{}/{}",
```

Comment:
We don't use the trace level elsewhere. Not sure if it is supported by all logging implementations.